py/parse.c

```diff
@@ -1073,12 +1080,21 @@ mp_parse_tree_t mp_parse(mp_lexer_t *lex, mp_parse_input_kind_t input_kind) {
             default: {
                 assert((rule_act & RULE_ACT_KIND_MASK) == RULE_ACT_LIST);
 
+                if (rule_id == RULE_arglist && !MP_PARSE_IS_PARSING_ARGLIST(&parser)) {
+                    parser.cur_view |= MP_PARSE_PARSING_ARGLIST;
```

Comment:
Instead of maintaining this bit, did you consider peeking into the rule stack?  This information should already be there, you just need to do something like `parser->rule_stack[parser->rule_stack_top - 1].rule_id == RULE_arglist`.

Reply:
I actually did, and I think the correct code to do this would be `parser->rule_stack[parser->rule_stack_top - 15].rule_id == RULE_arglist`, where 15 is that distance between the `RULE_arglist` and `RULE_atom_paren` that would guarantee that the second is a child of the first. My concern with doing this is that it creates a dependence on the grammar file, i.e. if someone makes a change to it (adding a rule), this would silently break. I went with the bit solution to hopefully add some robustness.

Reply:
Does the approach her work with nested function calls, eg `f(f(), (a)=2)`?

Reply:
Yep. On this input code, after parsing through the first argument `f()`, the parser will revisit the `arglist` rule (popping it from the rule stack) before recursing down through the `(a) = 2` argument. Added a test a did a quick run just now to confirm.